### ISSUE TITLE
Use with LastPass freezes app 10+ seconds

### ISSUE THREAD

**Issue by kevinbeal** [LogRocket/logrocket#9]
There is a function within the lastpass chrome extension that is intercepting XHR payloads (and maybe other things) and calling `JSON.stringify()` before doing something else with it.

Something within LogRocket is producing `UInt8Array`s with millions of numbers in it. This array gets passed to the above LastPass function. Stringifying millions of numbers is very slow.

**Comment by davidsneighbour** [comment]
https://lastpass.com/supportticket.php < wouldn't that be the right place to go?

**Comment by kevinbeal** [comment]
I did report the issue to them as well.

It's not so clear to me that it is their issue though. What in the world is LogRocket storing millions of numbers in a single array for? I'm shocked this is the first issue I've come across regarding it.

**Comment by davidsneighbour** [comment]
Those are probably the points that a mouse visited on the screen. I think that when you are logged into a site (as "defined" by the domain/subdomain you are on) then LastPass should be going into some kind of background mode. I often have issues with this plugin on ajax/js-heavy pages.

**Comment by kevinbeal** [comment]
LastPass may or may not be doing anything with this. The ticket was passed along to their dev team. They probably won't stop monkey-patching XHR.send(). It would be nice if they just ignored UInt8Array request bodies, but I doubt they will do that.

I don't have a solution for my users who use LastPass. Either I have to stop using LogRocket, or they have to put up with a terrible user experience.

It would be nice if LastPass went into some sort of background mode, but I don't think there exists a generalized way of detecting whether or not someone is logged into a site.

**Comment by govindrai** [comment]
@kevinbeal Were you able to figure out a solution to this? We are having the same issue, We hang for about 10-15 seconds when both lastpass and logrocket are running. @davidsneighbour anything you guys can do to communicate with last pass regarding this incompatability? 

Thanks!
Govind

**Comment by kevinbeal** [comment]
> @kevinbeal Were you able to figure out a solution to this?

No. LastPass is totally unresponsive. On a personal note, they've never addressed any issues I've sent them. I suspect they have no idea how to run a business and are coasting off of their initial success. I could be biased though since I lost thousands of dollars due to a lost password I assumed wrongly they would save.

**Comment by kevinbeal** [comment]
It occurs to me now that you could patch the `.stringify()` method to catch this case and just return an empty string or throw an error or something.

**Comment by govindrai** [comment]
@kevinbeal thanks! For the time being we are just turning off log rocket whenever last pass's monkey patching is present.  Btw how did you realize that it was XHR.send that was monkey patched? This thread was super helpful, we wouldn't have realized it was XHR.send that was causing issues without it.

**Comment by kevinbeal** [comment]
> Btw how did you realize that it was XHR.send that was monkey patched? 

I’m glad it was helpful!

I found it by waiting for the freeze to occur and hitting the dev tools pause button under Sources as fast as I could. The trace had the xhr patch from LastPass in it. I prettified the source at that location also using the dev tools. I saw the stringify call in there and put a breakpoint, again via dev tools. I got the app to freeze again and it immediately arrived at my breakpoint. I stepped over the function call with the debugger and it froze for a long time before it reached that next line in the source.

That’s my memory of the situation two years ago. I’m likely to have a detail wrong.

**Comment by govindrai** [comment]
Thanks for taking the time to respond, I super appreciate that explanation!! That's a good debugging tip for situations like this in the future, I'll keep it mind! Have a wonderful weekend! 😄

**Comment by arthur5005** [comment]
@govindrai can you share how you detect the LastPass monkey patch? We're having the same conflict and we want to make sure we don't load LogRocket in this case.

**Comment by govindrai** [comment]
@arthur5005 We added the below script in our html's `<head>` section.

```html
        <script>
                // returns true if a function has not been monkey-patched
                const isNative = fn => !!fn.toString().match(/\[native code\]/);
                
                // disables logrocket if send method has been monkey-patched (logrocket-lastpass compability issue)
                if (isNative(XMLHttpRequest.prototype.send)) {
                    var script = document.createElement('script');
                    script.type = 'text/javascript';
                    script.src = 'https://cdn.logrocket.io/LogRocket.min.js';
                    script.crossOrigin = 'anonymous'
                    document.getElementsByTagName('head')[0].appendChild(script);
                } else {
                    console.log("Disabling Log Rocket as XMLHttpRequest.prototype.send has been monkey-patched");
                }
       </script>
```

**Comment by coreymaher** [comment]
The best way I found to get around this issue without disabling logrocket is to use a response sanitizer (https://docs.logrocket.com/reference/network#sanitize-a-network-response).
In my case I had a large binary payload that was returned by a request which logrocket would capture and then lastpass would attempt to JSON encode. I added a response sanitizer that filtered out that specific large response and now the log rocket requests are small enough that lastpass doesn't freeze the browser.
